### PR TITLE
Partial C4CollectionDocObserver Tests

### PR DESCRIPTION
**Pull request by LaurenNguyen14**

### CODE REVIEW

File: common/main/java/com/couchbase/lite/internal/core/C4CollectionDocObserver.java [hunk 1]
```diff
@@ -21,11 +21,11 @@ public interface NativeImpl {
     //-------------------------------------------------------------------------
 
     @NonNull
-    private static final NativeImpl NATIVE_IMPL = new NativeC4CollectionDocObserver();
+    @VisibleForTesting
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/core/C4CollectionDocObserver.java:24]: I don't think there is any need for this.  You should never need to assign `nativeImpl` unless it is used in the `callback` method.  I don't think it is.

File: common/main/java/com/couchbase/lite/internal/core/C4CollectionDocObserver.java [hunk 2]
```diff
@@ -38,17 +38,17 @@ static void callback(long peer, @Nullable String docID, long sequence) {
 
         final C4CollectionDocObserver observer = BOUND_OBSERVERS.getBinding(peer);
         if (observer == null) { return; }
-
-        observer.listener.run();
+        observer.docChanged();
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/core/C4CollectionDocObserver.java:41]: I think you can improve your tests without changing this.  See comments below.

File: common/main/java/com/couchbase/lite/internal/core/C4CollectionDocObserver.java [hunk 3]
```diff
@@ -58,7 +58,10 @@ static C4CollectionDocObserver newObserver(
         long c4Coll,
         @NonNull String id,
         @NonNull Runnable listener) {
-        final C4CollectionDocObserver observer = new C4CollectionDocObserver(impl, impl.nCreate(c4Coll, id), listener);
+        final C4CollectionDocObserver observer = new C4CollectionDocObserver(
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/core/C4CollectionDocObserver.java:61]: Very nice.

File: common/main/java/com/couchbase/lite/internal/core/C4CollectionDocObserver.java [hunk 4]
```diff
@@ -76,7 +79,11 @@ static C4CollectionDocObserver newObserver(
     // Constructor
     //-------------------------------------------------------------------------
 
-    private C4CollectionDocObserver(@NonNull NativeImpl impl, long collection, @NonNull Runnable listener) {
+
+    private C4CollectionDocObserver(
+        @NonNull C4CollectionDocObserver.NativeImpl impl,
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/core/C4CollectionDocObserver.java:84]: FWIW, you don't need to qualify NativeImpl.  There's only one visible.

File: common/main/java/com/couchbase/lite/internal/core/C4CollectionDocObserver.java [hunk 5]
```diff
@@ -85,6 +92,9 @@ private C4CollectionDocObserver(@NonNull NativeImpl impl, long collection, @NonN
     @Override
     public void close() { closePeer(null); }
 
+    @VisibleForTesting
```
Comment (bmeike) [inline common/main/java/com/couchbase/lite/internal/core/C4CollectionDocObserver.java:95]: Yeah... See notes below.

File: common/test/java/com/couchbase/lite/internal/core/C4BaseTest.java [hunk 6]
```diff
@@ -115,6 +118,30 @@ protected void createRev(C4Database db, String docID, String revID, byte[] body)
         createRev(db, docID, revID, body, 0);
     }
 
+
+    // This method is a mock for c4Collection_putDoc. It checks whether a document is updated with a new revision id,
```
Comment (bmeike) [inline common/test/java/com/couchbase/lite/internal/core/C4BaseTest.java:122]: Cool.  Should we remove this when c4Collection_putDoc goes in?  ... or will it still be useful for testing?

File: common/test/java/com/couchbase/lite/internal/core/C4CollectionDocObserverTest.kt [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+import com.couchbase.lite.AbstractIndex
+import junit.framework.Assert.assertEquals
+import junit.framework.Assert.assertNotNull
+import org.junit.Ignore
+import org.junit.Test
+import java.util.concurrent.atomic.AtomicInteger
+
+class C4CollectionDocObserverTest : C4BaseTest() {
+    private val mockCollectionDocObserver = object : C4CollectionDocObserver.NativeImpl {
+        override fun nCreate(coll: Long, docId: String?): Long = 0xd8597341L
```
Comment (bmeike) [inline common/test/java/com/couchbase/lite/internal/core/C4CollectionDocObserverTest.kt:12]: Awesome!  Really nice.

File: common/test/java/com/couchbase/lite/internal/core/C4CollectionDocObserverTest.kt [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        override fun nFree(peer: Long) = Unit
+    }
+
+    @Test
+    fun testCreateC4CollectionDocObserver() {
+        val impl = C4CollectionDocObserver.nativeImpl
```
Comment (bmeike) [inline common/test/java/com/couchbase/lite/internal/core/C4CollectionDocObserverTest.kt:18]: I don't believe there is any need for this.  This is what the `create` that takes a native impl as parameter, is for.

**Review by bmeike — CHANGES_REQUESTED** [review]
This is good stuff, Lauren.  Nice.
There are two changes I'd like you to make:
1) I don't think there is ever a need to assign nativeImpl.  Please make it private and final.
2) I think there is a way to improve `testDocumentChanged`.  Don't change the implementations of `callback` or the Observer class, and call `callback` from the test.

File: common/test/java/com/couchbase/lite/internal/core/C4BaseTest.java (hunk 6 above)
Comment (LaurenNguyen14) [reply common/test/java/com/couchbase/lite/internal/core/C4BaseTest.java:122]: We can remove this when putDoc goes in

Comment (bmeike) [reply common/test/java/com/couchbase/lite/internal/core/C4BaseTest.java:122]: Great... Stick that in a comment, I think....

**Review by bmeike — APPROVED** [review]
That's great.  Push it!

**Pull request merged**